File: #47_Permutations_II/Solution.java
import java.util.*;

public class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        if (nums == null || nums.length == 0) return new LinkedList();
        Arrays.sort(nums);
        List<Integer> numList = new LinkedList();
        for (int n : nums) numList.add(n);
        return backtrack(numList);
    }

    private List<List<Integer>> backtrack(List<Integer> list) {
    	List<List<Integer>> ret = new LinkedList();
    	if (list.size() == 1) {
    		List<Integer> newList = new LinkedList(list);
    		ret.add(newList);
    		return ret;
    	}
    	for (int i = 0; i < list.size(); i++) {
    		if (i > 0 && list.get(i) == list.get(i - 1)) continue;
    		int tmp = list.remove(i);
    		List<List<Integer>> bt = backtrack(list);
    		for (List<Integer> ls : bt) {
    			ls.add(0, tmp);
    			ret.add(ls);
    		}
    		list.add(i, tmp);
    	}
    	return ret;
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	int[] t = {1, 2, 1};
    	List<List<Integer>> result = sol.permuteUnique(t);
    	for (List<Integer> ls : result) {
    		for (Integer it : ls) {
    			System.out.print(it + " ");
    		}
    		System.out.println();
    	}
    }
}

File: #47_Permutations_II/Solution.py
class Solution:
	# @param {integer[]} nums
	# @return {integer[][]}
	def permuteUnique(self, nums):
		if not nums:
			return []
		if len(nums) < 2:
			return [[nums[0]]]
		result = []
		hset = set()
		for x in range(len(nums)):
			if nums[x] in hset:
				continue
			hset.add(nums[x])
			self.swap(nums, 0, x)
			for comb in self.permuteUnique(nums[1:]):
				comb.append(nums[0])
				result.append(comb)
			self.swap(nums, 0, x)
		return result

	def swap(self, nums, i, j):
		if i == j:
			return
		tmp = nums[i]
		nums[i] = nums[j]
		nums[j] = tmp

sol = Solution()
print sol.permuteUnique([1, 2, 1])